src/components/NostoImage.ts

```diff
@@ -0,0 +1,65 @@
+import { customElement } from "./decorators"
+import { assertRequired } from "@/utils"
+import { transformUrl } from "unpic"
+import { transformBigcommerce } from "./NostoImge/bigcommerceHandler"
+
+@customElement("nosto-image")
+export class NostoImage extends HTMLElement {
+  static attributes = {
+    src: String,
+    width: String,
+    height: String
+  }
+
+  src!: string
+  width!: string
+  height!: string
+  transformer!: typeof transformUrl | typeof transformBigcommerce
+
+  connectedCallback() {
+    assertRequired(this, "src", "width", "height")
+    const { src, width, height } = this
+
+    this.transformer = this.getProvider() === "bigcommerce" ? transformBigcommerce : transformUrl
+    const transformedUrl = this.transformer({
+      url: src,
+      width: parseInt(width, 10),
+      height: parseInt(height, 10)
+    })
+
+    const imageHtml = `
+      <img 
+        src="${transformedUrl}" 
+        width="${width}" 
+        height="${height}" 
+        sizes="(max-width: 600px) 100vw, (max-width: 1200px) 50vw, 33vw"
+        srcset="${this.generateSrcset()}"
+        loading="lazy">
+    `
+    this.innerHTML = imageHtml
+  }
+
+  generateSrcset(): string {
+    const widths = [300, 600, 900, 1200]
+    return widths
+      .map(width => {
+        const url = this.transformer({
+          url: this.src,
+          width,
+          height: Math.round((width / 4) * 3)
```

Comment:
Does this retain the image aspect ratio? And what's the purpose of this calculation?